### ISSUE TITLE
lowercase language tag in URL

### ISSUE THREAD

**Issue by reeLx** [contao/contao#862]
Is it possible to write the language tag in the url in lowercase letters?
For example: Our customer doesn't want the url named like this

> https://www.domain.com/es-AR/

but instead like this

> https://www.domain.com/es-ar/

we tried to achieve that, by just renaming the tag into lowercase-letters in the changelanguage extension but that doesn't take any effect.

Greets
Marvin

**Comment by fritzmg** [comment]
I agree this should be possible. "In the wild" I always ever see it in lower case in URLs. e.g.

* https://www.microsoft.com/de-at/
* https://www.microsoft.com/de-de/
* https://www.microsoft.com/en-us/
* ...

**Comment by ausi** [comment]
I think this is related to https://github.com/contao/core-bundle/issues/1579#issuecomment-536687962

If we support every language tag or locale the *Intl* extension can parse the lowercase versions should work fine IMO because `Locale::parseLocale('de-at') === Locale::parseLocale('de_AT')`.

When implementing this feature we should look carefully where we should use the language tag as it was set in the backend – like in the URL – and where we want to use the canonical form (`Locale::canonicalize()`) – like for the translation files.

**Comment by leofeyer** [comment]
The intl extension seems to care neither about lowercase or uppercase nor locale or language ID:

```php
Locale::parseLocale('de-at'); // language: de, region: AT
Locale::parseLocale('de-AT'); // language: de, region: AT
Locale::parseLocale('de_at'); // language: de, region: AT
Locale::parseLocale('de_AT'); // language: de, region: AT
Locale::parseLocale('DE-AT'); // language: de, region: AT
Locale::parseLocale('DE-at'); // language: de, region: AT
```

See https://3v4l.org/EStaH

@contao/developers Do we want to do the same in Contao or do we want to support `xx-xx` and `xx-XX` only?

**Comment by leofeyer** [comment]
LOL, it seems the intl extension does not validate too much:

```php
Locale::parseLocale('23'); // language: 23
Locale::parseLocale(true); // language: 1
```

**Comment by Toflar** [comment]
Yes, the `parseLocale()` method doesn't validate anything. What are you trying to achieve?

**Comment by Toflar** [comment]
I guess what you're looking for is this: https://3v4l.org/B3W87

**Comment by Toflar** [comment]
Canonicalize would be setting `true` as third argument on `lookup()`.

**Comment by aschempp** [comment]
I don't think we need to validate anything. If the user uses `23` as their page language, so be it… it just won't load any matching language files (except obviously if a filter `languages/23/…` exists.

**Comment by ausi** [comment]
I agree with @aschempp we should not validate/lookup the locale of a root page against a list of known locales.

But we should use `Locale::canonicalize()` for things like access to language files. `Locale::canonicalize('23') === '23'`, `Locale::canonicalize('DEU') === 'de'`.

> Do we want to do the same in Contao or do we want to support `xx-xx` and `xx-XX` only?

I think we could support any alpa-numeric string including `-` and `_`. Maybe with a length limit?

**Comment by aschempp** [comment]
A correct language for a traditional chinese website would be `cn-Hant`, because that means China with Traditional Characters (https://en.wikipedia.org/wiki/Traditional_Chinese_characters). People currently use `cn-TW` for Taiwan, because that's where traditional writing is mostly used. On the same note, one could use `de-Cyrl` to write german in cyrillic letters (which is totally possible).

So again, browser detection will not work (though it will work for Symfony language files!), but we can't limit people. Again, see https://github.com/contao/contao/issues/6358 for extensive infos.

I suggest to move this to Contao 4.10 and review the (whole) implementation.

**Comment by aschempp** [comment]
FYI this has been closed because in Contao 4.10 the URL prefix can be configured independently of the page language 😎
